### PR TITLE
Changed async test structure & small cleanups & fast check integration

### PR DESCRIPTION
**Pull request by Emulebest**
The fix to #229 without messing in parts of sorting is implemented in #230

### CODE REVIEW

File: papiea-engine/__tests__/api_docs.test.ts [hunk 1]
```diff
@@ -62,21 +62,21 @@ describe("API Docs Tests", () => {
...
+        expect.hasAssertions();
         try {
             const apiDoc = await apiDocsGenerator.getApiDocs();
             const apiDocJson = JSON.stringify(apiDoc);
             writeFileSync("api-docs.json", apiDocJson);
             const api = await validate("api-docs.json");
             expect(api.info.title).toEqual("Swagger Papiea");
-            done();
         } catch (err) {
-            done.fail(err);
```
Comment (adolgarev) [inline papiea-engine/__tests__/api_docs.test.ts]: I believe you should throw here an error or remove catch() block (leave try-finally). Do not like expect.hasAssertions() approach

File: papiea-engine/__tests__/database_mongo.test.ts [hunk 2]
```diff
@@ -72,50 +77,50 @@ describe("MongoDb tests", () => {
...
+            expect(res).not.toBeNull();
+            if (res === null) {
+                throw new Error("Entity without spec");
+            }
+            const [metadata, spec] = res;
+            expect(metadata.uuid).toEqual(entity_ref.uuid);
+            expect(metadata.created_at).not.toBeNull();
+            expect(metadata.deleted_at).toBeFalsy();
+            expect(spec.a).toEqual("A1");
+        } catch (e) {
```
Comment (adolgarev) [inline papiea-engine/__tests__/database_mongo.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/database_mongo.test.ts [hunk 3]
```diff
@@ -124,141 +129,143 @@ describe("MongoDb tests", () => {
...
-            return;
+        try {
+            expect(res).not.toBeNull();
+            if (res === null) {
+                throw new Error("Entity without status");
+            }
+            const [metadata, status] = res;
+            expect(metadata.uuid).toEqual(entity_ref.uuid);
+            expect(status.a).toEqual("A1");
+        } catch (e) {
```
Comment (adolgarev) [inline papiea-engine/__tests__/database_mongo.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/api_docs.test.ts [hunk 4]
```diff
@@ -91,12 +91,12 @@ describe("API Docs Tests", () => {
             expect(Object.keys(kindEntityPath)).toContain("get");
             expect(Object.keys(kindEntityPath)).toContain("delete");
             expect(Object.keys(kindEntityPath)).toContain("put");
-            done();
         } catch (err) {
-            done.fail(err);
+            throw err;
```
Comment (adolgarev) [inline papiea-engine/__tests__/api_docs.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/api_docs.test.ts [hunk 5]
```diff
@@ -128,9 +128,8 @@ describe("API Docs Tests", () => {
                     }
                 }
             });
-            done();
         } catch (err) {
-            done.fail(err);
+            throw err;
```
Comment (adolgarev) [inline papiea-engine/__tests__/api_docs.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/api_docs.test.ts [hunk 6]
```diff
@@ -175,14 +175,14 @@ describe("API docs test entity", () => {
                 .content["application/json"]
                 .schema["$ref"]).toEqual(`#/components/schemas/SumOutput`);
 
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/api_docs.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/api_docs.test.ts [hunk 7]
```diff
@@ -225,13 +225,13 @@ describe("API docs test entity", () => {
                     }
                 }
             });
-            done();
         } catch (err) {
-            done.fail(err);
+            throw err;
```
Comment (adolgarev) [inline papiea-engine/__tests__/api_docs.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/api_docs.test.ts [hunk 8]
```diff
@@ -266,10 +266,8 @@ describe("API docs test entity", () => {
                 .responses["200"]
                 .content["application/json"]
                 .schema["$ref"]).toEqual(`#/components/schemas/Nothing`);
-
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/api_docs.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/database_mongo.test.ts [hunk 9]
```diff
@@ -124,141 +129,143 @@ describe("MongoDb tests", () => {
...
+        try {
+            expect(res).not.toBeNull();
+            if (res === null) {
+                throw new Error("Entity without status");
+            }
+            const [metadata, status] = res;
+            expect(metadata.uuid).toEqual(entity_ref.uuid);
+            expect(status.a).toEqual("A1");
+            expect(status.b).toEqual("A3");
+        } catch (e) {
```
Comment (adolgarev) [inline papiea-engine/__tests__/database_mongo.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 10]
```diff
@@ -60,13 +62,12 @@ describe("Entity API tests", () => {
             expect(spec).not.toBeUndefined();
             entity_metadata = metadata;
             entity_spec = spec;
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 11]
```diff
@@ -75,40 +76,38 @@ describe("Entity API tests", () => {
...
         expect.assertions(2);
         try {
             const res = await entityApi.get(`/${providerPrefix}/${providerVersion}/${kind_name}/${entity_metadata.uuid}`);
             expect(res.data.spec).toEqual(entity_spec);
             expect(res.data.status).toEqual(entity_spec);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 12]
```diff
@@ -125,13 +124,14 @@ describe("Entity API tests", () => {
                 }
             });
             expect(res.data.results.length).toBe(0);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 13]
```diff
@@ -143,13 +143,14 @@ describe("Entity API tests", () => {
             const entity = res.data.results[0];
             expect(entity.spec.x).toEqual(10);
             expect(entity.status).toEqual(entity.spec);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 14]
```diff
@@ -178,13 +179,14 @@ describe("Entity API tests", () => {
                 }
             });
             expect(res.data.results.length).toBe(0);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 15]
```diff
@@ -203,13 +205,14 @@ describe("Entity API tests", () => {
                 expect(entity.status.x).toEqual(10);
                 expect(entity.status.y).toEqual(11);
             });
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 16]
```diff
@@ -309,14 +312,14 @@ describe("Entity API tests", () => {
                 expect(entity.status.v.e).toEqual(12);
             });
 
-            done();
+
         } catch (e) {
-            console.log(e.response.data.errors);
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 17]
```diff
@@ -327,13 +330,13 @@ describe("Entity API tests", () => {
         try {
             const res = await entityApi.get(`${providerPrefix}/${providerVersion}/${kind_name}?${stringify(spec_query)}`);
             expect(res.data.results.length).toBeGreaterThanOrEqual(1);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 18]
```diff
@@ -349,13 +352,13 @@ describe("Entity API tests", () => {
             res = await entityApi.get(`/${providerPrefix}/${providerVersion}/${kind_name}/${entity_metadata.uuid}`);
             expect(res.data.spec.x).toEqual(20);
             expect(res.data.status.x).toEqual(20);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 19]
```diff
@@ -414,28 +417,29 @@ describe("Entity API tests", () => {
             expect(metadata.uuid).toEqual(entity_uuid);
             const res = await entityApi.get(`/${providerPrefix}/${providerVersion}/${kind_name}/${entity_uuid}`);
             expect(res.data.metadata.uuid).toEqual(entity_uuid);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 20]
```diff
@@ -414,28 +417,29 @@ describe("Entity API tests", () => {
...
     });
 
-    test("Delete entity", async (done) => {
+    test("Delete entity", async () => {
+        expect.assertions(1);
         try {
             await entityApi.delete(`/${providerPrefix}/${providerVersion}/${kind_name}/${entity_metadata.uuid}`);
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 21]
```diff
@@ -461,9 +465,9 @@ describe("Entity API tests", () => {
                 expect(res.data.results[0].spec).toEqual(spec);
                 expect(res.data.results[0].status).toEqual(spec);
             });
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 22]
```diff
@@ -526,25 +531,26 @@ describe("Entity API with metadata extension tests", () => {
             });
             entity_metadata = metadata;
             entity_spec = spec;
-            done();
+            expect(spec).toBeDefined();
+            expect(metadata).toBeDefined();
         } catch (err) {
-            done.fail(err);
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 23]
```diff
@@ -526,25 +531,26 @@ describe("Entity API with metadata extension tests", () => {
...
         expect.assertions(2);
         try {
             const res = await entityApi.get(`/${providerPrefix}/${providerVersion}/${kind_name}/${entity_metadata.uuid}`);
             expect(res.data.spec).toEqual(entity_spec);
             expect(res.data.status).toEqual(entity_spec);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 24]
```diff
@@ -576,31 +581,32 @@ describe("Entity API with metadata extension tests", () => {
                 }
             });
             expect(res.data.results.length).toBe(1);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 25]
```diff
@@ -576,31 +581,32 @@ describe("Entity API with metadata extension tests", () => {
...
             expect(res.status).toEqual(400);
             expect(res.data.errors.length).toEqual(1);
             expect(res.data.errors[0]).toEqual("Metadata extension is not specified");
-            done();
+
         }
     });
 });
 
+// TODO: rewrite the whole suite with fast check
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: what for?

File: papiea-engine/__tests__/entity_api.test.ts [hunk 26]
```diff
@@ -635,13 +641,13 @@ describe("Pagination tests", () => {
             const entityResponses: any[] = await Promise.all(entityPromises);
             uuids = entityResponses.map(entityResp => entityResp.data.metadata.uuid);
             expect(entityResponses.length).toBe(70);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 27]
```diff
@@ -651,13 +657,14 @@ describe("Pagination tests", () => {
             });
             expect(res.data.results.length).toBe(30);
             expect(res.data.entity_count).toBe(70);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 28]
```diff
@@ -667,13 +674,14 @@ describe("Pagination tests", () => {
             });
             expect(res.data.results.length).toBe(10);
             expect(res.data.entity_count).toBe(70);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 29]
```diff
@@ -683,13 +691,13 @@ describe("Pagination tests", () => {
             });
             expect(res.data.results.length).toBe(30);
             expect(res.data.entity_count).toBe(70);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 30]
```diff
@@ -699,82 +707,82 @@ describe("Pagination tests", () => {
             });
             expect(res.data.results.length).toBe(20);
             expect(res.data.entity_count).toBe(70);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 31]
```diff
@@ -699,82 +707,82 @@ describe("Pagination tests", () => {
...
         const deletePromises: Promise<any>[] = [];
         try {
             uuids.forEach(uuid => {
                 deletePromises.push(entityApi.delete(`/${providerPrefix}/${providerVersion}/${kind_name}/${uuid}`));
             });
             await Promise.all(deletePromises);
         } catch (e) {
-            done.fail(e);
-            return;
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 32]
```diff
@@ -784,15 +792,15 @@ describe("Pagination tests", () => {
                 }
             });
             expect(res.data.results.length).toBe(0);
-            done();
+
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 33]
```diff
@@ -838,18 +846,84 @@ describe("Pagination tests", () => {
...
                 spec: {
                     y: 11
                 }
             });
-            console.log(data.results[0]);
             expect(data.results[0].spec.x).toBe(0);
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 34]
```diff
@@ -838,18 +846,84 @@ describe("Pagination tests", () => {
...
+        expect.assertions(1);
+        try {
+            const { data } = await entityApi.post(`${providerPrefix}/${providerVersion}/${kind_name}/filter?sort=spec.x:desc`, {
+                spec: {
+                    y: 11
+                }
+            });
+            expect(data.results[0].spec.x).toBe(69);
+        } catch (e) {
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api.test.ts [hunk 35]
```diff
@@ -838,18 +846,84 @@ describe("Pagination tests", () => {
...
+        expect.assertions(1);
+        try {
+            const { data } = await entityApi.post(`${providerPrefix}/${providerVersion}/${kind_name}/filter?sort=spec.y,spec.x`, {
+                spec: {
+                    y: 11
+                }
+            });
+            expect(data.results[0].spec.x).toBe(0);
+        } catch (e) {
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 36]
```diff
@@ -208,13 +209,13 @@ describe("Entity API auth tests", () => {
             expect(data.owner).toEqual("alice");
             expect(data.tenant).toEqual(tenant_uuid);
             expect(data.provider_prefix).toEqual(provider.prefix);
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 37]
```diff
@@ -236,13 +237,13 @@ describe("Entity API auth tests", () => {
             );
             expect(data.owner).toEqual("alice");
             expect(data.tenant).toEqual(tenant_uuid);
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 38]
```diff
@@ -269,57 +269,61 @@ describe("Entity API auth tests", () => {
             );
             expect(metadata).toEqual(entity_metadata);
             expect(spec).toEqual(entity_spec);
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 39]
```diff
@@ -269,57 +269,61 @@ describe("Entity API auth tests", () => {
...
+            const { data: { token } } = await providerApi.get(`/${ provider.prefix }/${ provider.version }/auth/login`);
+            await providerApiAdmin.post(`/${ provider.prefix }/${ provider.version }/auth`, {
+                policy: `p, alice, owner, ${ kind_name }, *, allow`
+            });
+            const result = await entityApi.post(`/${ provider.prefix }/${ provider.version }/${ kind_name }/${ entity_metadata.uuid }/procedure/moveX`, { input: 5 },
+                { headers: { 'Authorization': 'Bearer ' + token } }
+            );
+            expect(result);
+        } catch (e) {
+
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 40]
```diff
@@ -384,13 +388,12 @@ describe("Entity API auth tests", () => {
             );
             expect(metadata).toEqual(entity_metadata);
             expect(spec).toEqual(entity_spec);
-            done();
         } catch (e) {
-            done.fail(e);
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 41]
```diff
@@ -415,16 +418,17 @@ describe("Entity API auth tests", () => {
...
                 );
-                done.fail("Key created with another provider");
+                throw new Error("Key created with another provider");
             } catch (e) {
+                expect(e).toBeDefined();
             }
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 42]
```diff
@@ -464,16 +468,17 @@ describe("Entity API auth tests", () => {
...
             });
-            await entityApi.post(`/${provider.prefix}/${provider.version}/${kind_name}/${entity_metadata.uuid}/procedure/moveX`, { input: 5 },
+            const result = await entityApi.post(`/${provider.prefix}/${provider.version}/${kind_name}/${entity_metadata.uuid}/procedure/moveX`, { input: 5 },
                 { headers: { 'Authorization': 'Bearer ' + token } }
             );
-            done();
+            expect(result);
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 43]
```diff
@@ -521,16 +526,17 @@ describe("Entity API auth tests", () => {
...
             );
-            await entityApi.post(`/${provider.prefix}/${provider.version}/${kind_name}/procedure/computeGeolocation`, { input: "2" },
+            const result = await entityApi.post(`/${provider.prefix}/${provider.version}/${kind_name}/procedure/computeGeolocation`, { input: "2" },
                 { headers: { 'Authorization': 'Bearer ' + s2skey.key } }
             );
-            done();
+            expect(result).toBeDefined();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 44]
```diff
@@ -584,21 +590,22 @@ describe("Entity API auth tests", () => {
...
                     "b": 5
                 }
             },
                 { headers: { 'Authorization': 'Bearer ' + s2skey.key } }
             );
-            done();
+            expect(result).toBeDefined();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/procedures_api.test.ts [hunk 45]
```diff
@@ -189,15 +185,14 @@ describe("Procedures tests", () => {
                 }
             });
             expect(res.data).toBe(10);
-            done();
         } catch (e) {
-            console.log(e.response.data);
-            done.fail()
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/procedures_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_api.test.ts [hunk 46]
```diff
@@ -152,13 +151,13 @@ describe("Provider API tests", () => {
             });
             const { data: { policy } } = await providerApi.get(`/${provider.prefix}/${provider.version}`);
             expect(policy).toEqual(originalPolicy);
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_api.test.ts [hunk 47]
```diff
@@ -184,9 +183,8 @@ describe("Provider API tests", () => {
             expect(res.data.status.x).toEqual(10);
             expect(res.data.status.y).toEqual(20);
             expect(res.data.status.z).toEqual(111);
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_api.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_api_auth.test.ts [hunk 48]
```diff
@@ -58,13 +59,14 @@ describe("Provider API auth tests", () => {
             );
             expect(userInfo.provider_prefix).toEqual(provider.prefix);
             expect(userInfo.is_provider_admin).toBeTruthy();
-            done();
         } catch (e) {
-            done.fail(e);
+            expect(e).toBeUndefined();
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_api_auth.test.ts [hunk 49]
```diff
@@ -82,13 +84,13 @@ describe("Provider API auth tests", () => {
...
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_api_auth.test.ts [hunk 50]
```diff
@@ -404,13 +408,13 @@ describe("Provider API auth tests", () => {
             );
             expect(data.length).toEqual(1);
             expect(data[0].owner).toEqual("user1@provider");
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_api_auth.test.ts [hunk 51]
```diff
@@ -440,9 +444,8 @@ describe("Provider API auth tests", () => {
             );
             expect(data.length).toEqual(1);
             expect(data[0].owner).toEqual("admin123@provider");
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_api_auth.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_sdk.test.ts [hunk 52]
```diff
@@ -147,11 +142,11 @@ describe("Provider Sdk tests", () => {
                 sdk.server.close()
             } catch(e) {}
         } catch (e) {
-            done.fail(e)
+            expect(e).toBeUndefined();
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_sdk.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_sdk.test.ts [hunk 53]
```diff
@@ -166,14 +161,15 @@ describe("Provider Sdk tests", () => {
         });
         try {
             await sdk.register();
-            done();
         } catch (e) {
-            done.fail(e)
+            expect(e).toBeUndefined();
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_sdk.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_sdk.test.ts [hunk 54]
```diff
@@ -200,14 +196,14 @@ describe("Provider Sdk tests", () => {
             const updatedEntity: any = await axios.get(`${sdk.entity_url}/${sdk.provider.prefix}/${sdk.provider.version}/${kind_name}/${metadata.uuid}`);
             expect(updatedEntity.data.metadata.spec_version).toEqual(2);
             expect(updatedEntity.data.spec.x).toEqual(15);
-            done();
         } catch (e) {
-            done.fail(e);
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_sdk.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_sdk.test.ts [hunk 55]
```diff
@@ -285,15 +279,16 @@ describe("Provider Sdk tests", () => {
 
         try {
             await sdk.register();
-            done();
         } catch (e) {
-            done.fail(e)
+            expect(e).toBeUndefined();
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_sdk.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_sdk.test.ts [hunk 56]
```diff
@@ -321,15 +316,15 @@ describe("Provider Sdk tests", () => {
         try {
             const res: any = await axios.post(`${sdk.entity_url}/${sdk.provider.prefix}/${sdk.provider.version}/${kind_name}/procedure/computeGeolocation`, { input: "2" });
             expect(res.data).toBe("us.west.2");
-            done();
         } catch (e) {
-            done.fail(e)
+            expect(e).toBeUndefined();
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_sdk.test.ts]: do not need try-catch here

File: papiea-engine/__tests__/provider_sdk.test.ts [hunk 57]
```diff
@@ -394,15 +382,16 @@ describe("Provider Sdk tests", () => {
         try {
             const res: any = await axios.post(`${sdk.entity_url}/${sdk.provider.prefix}/${sdk.provider.version}/procedure/computeSum`, { input: { "a": 5, "b": 5 } });
             expect(res.data).toBe(10);
-            done();
         } catch (e) {
-            done.fail(e)
+            expect(e).toBeUndefined();
+            throw e;
```
Comment (adolgarev) [inline papiea-engine/__tests__/provider_sdk.test.ts]: do not need try-catch here

File: papiea-engine/package.json [hunk 58]
```diff
@@ -30,6 +30,7 @@
     "express": "^4.16.4",
     "express-async-handler": "^1.1.4",
     "express-swaggerize-ui": "^1.0.3",
+    "fast-check": "^1.16.0",
```
Comment (adolgarev) [inline papiea-engine/package.json]: what for?

**Comment by shlomi-nx** [comment]
@Emulebest Does this PR relates to any issue? If not, could you please create an issue describing what is solved here, and link to it appropriately?

File: papiea-engine/__tests__/database_mongo.test.ts [hunk 59]
```diff
@@ -15,22 +15,27 @@ declare var process: {
         MONGO_PORT: string
     }
 };
-const mongoHost = process.env.MONGO_HOST || 'mongo'
-const mongoPort = process.env.MONGO_PORT || '27017'
+const mongoHost = process.env.MONGO_HOST || 'mongo';
```
Comment (shlomi-nx) [inline papiea-engine/__tests__/database_mongo.test.ts]: I am personally not a big fan of ending lines with `;`

File: papiea-engine/package.json (hunk 58 above)
Comment (adolgarev) [reply papiea-engine/package.json]: not sure that we need this, anyways move to dev deps

Comment (adolgarev) [reply papiea-engine/package.json]: @Emulebest ^^^

**Pull request closed without merging**